### ISSUE TITLE
Remote transactions set PRESERVE_BNODE_IDS to true and break the expected behaviour of blank node uniqueness across imported files

### ISSUE THREAD

**Issue by avataar** [eclipse/rdf4j#3623]
### Current Behavior

Given a remote repository, a transaction and importing two different files that share a blank node identifier, for example:

```java
try (RepositoryConnection connection = repository.getConnection()) {
    connection.begin();
    connection.add(new StringReader("<urn:a> <urn:b> _:c ."), RDFFormat.TURTLE);
    connection.add(new StringReader("<urn:a> <urn:d> _:c ."), RDFFormat.TURTLE);
    connection.commit();
}
```

Both instances of `_:c` will be imported exactly as `_:c`.

### Expected Behavior

Blank nodes should be scoped within a given serialization and thus the two instances should resolve to two unique blank nodes. This is actually how RDF4J works with local transactions.

The issue was introduced in ancient Sesame times here: https://bitbucket.org/openrdf/sesame/commits/2246487cdf0966eb46451498e443d471a7291b62, and later the code was migrated to the class Transaction. I couldn't find any explanation as to why it was needed.

As a side note: using `connection.getParserConfig().set(BasicParserSettings.PRESERVE_BNODE_IDS, false)` on a remote connection doesn't seem to do anything. It should either work as expected (affecting the settings) or throw an exception that setting parser settings on a remote connection isn't possible.

### Version

3.7.4

### Are you interested in contributing a solution yourself?

Yes

**Comment by jeenbroekstra** [comment]
> As a side note: using connection.getParserConfig().set(BasicParserSettings.PRESERVE_BNODE_IDS, false) on a remote connection doesn't seem to do anything. It should either work as expected (affecting the settings) or throw an exception that setting parser settings on a remote connection isn't possible.

Setting parser config on a remote connection  does affect actual parser settings, but it's not the parser that's being used at the server end you're configuring. The client-side also has a parser (used for processing server responses), and _that_ is what you configure. There's currently no way to control the server's parser remotely via the API, though there's a feature request to allow that kind of thing (#446).

**Comment by jeenbroekstra** [comment]
> The issue was introduced in ancient Sesame times here: https://bitbucket.org/openrdf/sesame/commits/2246487cdf0966eb46451498e443d471a7291b62, and later the code was migrated to the class Transaction. I couldn't find any explanation as to why it was needed.

The reason this was done is _probably_ that we need to preserve blank node identity in remote operations when it's updates on statements already in the store we're processing, or when using blank nodes minted via the API. For example:

```java
try (RepositoryConnection connection = repository.getConnection()) {
    BNode node1 = Values.bnode("node1");
    connection.begin();
    connection.add(node1, RDFS.LABEL, Values.literal("a bnode"));
    connection.add(node1, RDF.TYPE, Values.iri("http://example.org/BNode"));
    connection.commit();
}
```

The above should always result in two statements using the same subject blank node - and this is why on remote transaction, preservation of blank nodes is used at the server end.

Your example case does look odd though: I would have expected the  client-side parser to deal with this. I'll try and reproduce on my end.

**Comment by jeenbroekstra** [comment]
It's odd - I can reproduce the problem you're seeing, but only when explicitly using a transaction. I think we may be doing some sort of bypass of client-side parsing in explicit transactions, causing a difference. 

We could tweak the client side to always handle parsing but of course that brings a substantial performance penalty with it. Alternatively, we can look into ways in which we can distinguish, server-side, the intent, and tweak the server-side parser config accordingly.

**Comment by jeenbroekstra** [comment]
I had a closer look and actually the fix that @DenitsaStoianova proposed in PR #3662 is the correct one: for some reason we override the preserve Bnode ID parameter (which has been communicated from client to server) by just hardcoding the underlying connection to always set this to `true` - this only happens when using an explicit transaction, which explains the difference in behavior I saw earlier. So all good!

**Comment by konstantinspetrov** [comment]
@jeenbroekstra could you please also include the fix in 4.0.0-M2?

**Comment by hmottestad** [comment]
It will be available in the next milestone release, 4.0.0-M3. Planned in about two weeks.
